soroban-env-host/observations.json

```diff
@@ -0,0 +1,1091 @@
+{
```

Comment:
This looks great, thanks! The only UX-side thing I would probably consider doing (probably not in this PR) is to output step diffs where possible (cpu and mem are trivial, storage and objects are possible in theory thanks to immutability). The reason is that diffs are easier to read and also are better for the cases when test is modified in the middle (we could of course try to avoid these, but I suspect we'll still need to do that from time to time).

Reply:
I'm not sure what you mean by step diffs but I should highlight two facts here:

1. I'm emitting a single JSON map and rendering each K:V entry in it manually to make it compact for reading; this means the structure of each line is lost on rendering and can't be recovered. it's just "a string" after rendering. so I can't diff the fields within that string.
2. I'm not recording anything else (eg. the actual content of the storage or whatever) from previous runs, _just_ this one file. so we can't diff the preimage (storage state, PRNG state, whatever) that went into each field either.

Reply:
By 'diffs' I mean the difference between the current step and the previous step. At least for CPU/mem that seems straightforward (so that instead of `step 0: cpu: 1000`, `step 1: cpu: 2500` you'd get `step 0: cpu_delta: 1000`, `step 1: cpu_delta: 1500`. E.g. for the host objects it should be enough to hash not the whole array, but only the slice of (`prev_step_object_count..curr_step_object_count`). For storage it should be enough to store the storage reference for every step and then diff it (we're maintaining all the storage states anyway, right?).

I just want to make it clear again, that this suggestion definitely shouldn't block this PR. I realize it might require a bit of refactoring and it's not that high priority. Just something I'd maybe take a look at, maybe some time after release.

Reply:
ah, I see. I was able to accomplish a minimal version of this -- omitting fields that don't change from one line to the next -- without much refactoring. I also hooked the vmcallerenv-based (direct) calls, which reveals quite a lot more traffic (also helps explain why so many host objects -- the env macros aren't using bulk object construction I guess).

let me know if the new file reads easily enough. I think I might also change to a separate observations file per test, the unified "observations.json" file is already 1.9MiB and it's going to get a lot bigger. Github has a 100MiB file size limit and diff certainly isn't going to like diffing files that large either.